File: JavaScript/DesignPatterns/RATELIMITING.js
'use strict';

class Request {
    constructor(clientId, timestamp = Date.now() / 1000) {
        this.clientId = clientId;
        this.timestamp = timestamp;
    }
}

class IRateLimitingStrategy {
    isAllowed(request) {
        throw new Error("Not implemented");
    }
}

class TokenBucketStrategy extends IRateLimitingStrategy {
    constructor(capacity, refillRatePerSec) {
        super();
        this.capacity = capacity;
        this.refillRatePerSec = refillRatePerSec;
        this.tokens = capacity;
        this.lastRefill = Date.now() / 1000;
    }

    _refill(now) {
        const elapsed = now - this.lastRefill;
        this.tokens = Math.min(this.capacity, this.tokens + elapsed * this.refillRatePerSec);
        this.lastRefill = now;
    }

    isAllowed(request) {
        this._refill(request.timestamp);
        if (this.tokens >= 1) {
            this.tokens -= 1;
            return true;
        }
        return false;
    }
}

class SlidingWindowStrategy extends IRateLimitingStrategy {
    constructor(windowSizeSec, maxRequests) {
        super();
        this.windowSizeSec = windowSizeSec;
        this.maxRequests = maxRequests;
        this.timestamps = [];
    }

    isAllowed(request) {
        const now = request.timestamp;
        while (this.timestamps.length > 0 && this.timestamps[0] <= now - this.windowSizeSec) {
            this.timestamps.shift();
        }
        if (this.timestamps.length < this.maxRequests) {
            this.timestamps.push(now);
            return true;
        }
        return false;
    }
}

class NotificationService {
    constructor() {
        this.strategy = null;
    }

    setStrategy(strategy) {
        this.strategy = strategy;
    }

    send(request) {
        if (this.strategy === null) {
            throw new Error("No rate limiting strategy configured");
        }
        if (this.strategy.isAllowed(request)) {
            console.log(`Notification sent for client ${request.clientId}`);
        } else {
            console.log(`Rate limit exceeded for client ${request.clientId}`);
        }
    }
}

function main() {
    const service = new NotificationService();

    service.setStrategy(new TokenBucketStrategy(3, 1));
    console.log("-- Token Bucket --");
    for (let i = 0; i < 5; i++) {
        service.send(new Request("user-1"));
    }

    service.setStrategy(new SlidingWindowStrategy(60, 2));
    console.log("-- Sliding Window --");
    for (let i = 0; i < 4; i++) {
        service.send(new Request("user-2"));
    }
}

main();

File: Golang/DesignPatterns/RATELIMITING.go
package main

import (
	"fmt"
	"time"
)

type Request struct {
	clientID  string
	timestamp float64
}

func NewRequest(clientID string) Request {
	return Request{clientID: clientID, timestamp: nowSeconds()}
}

func nowSeconds() float64 {
	return float64(time.Now().UnixNano()) / 1e9
}

type RateLimitingStrategy interface {
	IsAllowed(request Request) bool
}

type TokenBucketStrategy struct {
	capacity         float64
	refillRatePerSec float64
	tokens           float64
	lastRefill       float64
}

func NewTokenBucketStrategy(capacity int, refillRatePerSec float64) *TokenBucketStrategy {
	return &TokenBucketStrategy{
		capacity:         float64(capacity),
		refillRatePerSec: refillRatePerSec,
		tokens:           float64(capacity),
		lastRefill:       nowSeconds(),
	}
}

func (s *TokenBucketStrategy) refill(now float64) {
	elapsed := now - s.lastRefill
	s.tokens = min(s.capacity, s.tokens+elapsed*s.refillRatePerSec)
	s.lastRefill = now
}

func (s *TokenBucketStrategy) IsAllowed(request Request) bool {
	s.refill(request.timestamp)
	if s.tokens >= 1 {
		s.tokens--
		return true
	}
	return false
}

type SlidingWindowStrategy struct {
	windowSizeSec float64
	maxRequests   int
	timestamps    []float64
}

func NewSlidingWindowStrategy(windowSizeSec float64, maxRequests int) *SlidingWindowStrategy {
	return &SlidingWindowStrategy{windowSizeSec: windowSizeSec, maxRequests: maxRequests}
}

func (s *SlidingWindowStrategy) IsAllowed(request Request) bool {
	now := request.timestamp
	for len(s.timestamps) > 0 && s.timestamps[0] <= now-s.windowSizeSec {
		s.timestamps = s.timestamps[1:]
	}
	if len(s.timestamps) < s.maxRequests {
		s.timestamps = append(s.timestamps, now)
		return true
	}
	return false
}

type NotificationService struct {
	strategy RateLimitingStrategy
}

func (n *NotificationService) SetStrategy(strategy RateLimitingStrategy) {
	n.strategy = strategy
}

func (n *NotificationService) Send(request Request) {
	if n.strategy == nil {
		panic("No rate limiting strategy configured")
	}
	if n.strategy.IsAllowed(request) {
		fmt.Printf("Notification sent for client %s\n", request.clientID)
	} else {
		fmt.Printf("Rate limit exceeded for client %s\n", request.clientID)
	}
}

func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func main() {
	service := &NotificationService{}

	service.SetStrategy(NewTokenBucketStrategy(3, 1))
	fmt.Println("-- Token Bucket --")
	for i := 0; i < 5; i++ {
		service.Send(NewRequest("user-1"))
	}

	service.SetStrategy(NewSlidingWindowStrategy(60, 2))
	fmt.Println("-- Sliding Window --")
	for i := 0; i < 4; i++ {
		service.Send(NewRequest("user-2"))
	}
}
